### PR TITLE
[SPARK-19447] Make Range operator generate "recordsRead" metric

### PR DESCRIPTION
**Pull request by ala**
## What changes were proposed in this pull request?

The Range was modified to produce "recordsRead" metric instead of "generated rows". The tests were updated and partially moved to SQLMetricsSuite.

## How was this patch tested?

Unit tests.

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala [hunk 1]
```diff
@@ -309,4 +314,84 @@ class SQLMetricsSuite extends SparkFunSuite with SharedSQLContext {
...
+              }
+            case _ => // Ignore.
+          }
+        }
+      }
+      res.sumMaxOutputRows += maxOutputRows
+    }
+  }
+
+  def run(df: DataFrame): List[(Long, Long, Long)] = {
```
Comment (rxin) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala]: document what hte long long long are for?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala [hunk 2]
```diff
@@ -309,4 +314,84 @@ class SQLMetricsSuite extends SparkFunSuite with SharedSQLContext {
...
+        var shuffleRecordsRead: Long = 0L,
+        var sumMaxOutputRows: Long = 0L)
+
+    private[this] var stageIdToMetricsResult = HashMap.empty[Int, MetricsResult]
+
+    def reset(): Unit = {
+      stageIdToMetricsResult = HashMap.empty[Int, MetricsResult]
+    }
+
+    def getResults(): List[(Long, Long, Long)] = {
```
Comment (rxin) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala]: here too long long long

**Comment by rxin** [comment]
cc @hvanhovell if you have a min to review this ...

File: sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala [hunk 3]
```diff
@@ -309,4 +314,94 @@ class SQLMetricsSuite extends SparkFunSuite with SharedSQLContext {
...
+        spark.range(0, 30).repartition(3).crossJoin(sql("select * from pqS")).repartition(2).toDF()
+      )
+      assert(res3 === (10L, 0L, 10L) :: (30L, 0L, 30L) :: (0L, 30L, 300L) :: (0L, 300L, 0L) :: Nil)
+    }
+  }
+}
+
+object InputOutputMetricsHelper {
+   private class InputOutputMetricsListener extends SparkListener {
+    private case class MetricsResult(
```
Comment (hvanhovell) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala]: Nit: add space

File: sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala [hunk 4]
```diff
@@ -309,4 +314,94 @@ class SQLMetricsSuite extends SparkFunSuite with SharedSQLContext {
...
+        var recordsRead: Long = 0L,
+        var shuffleRecordsRead: Long = 0L,
+        var sumMaxOutputRows: Long = 0L)
+
+    private[this] var stageIdToMetricsResult = HashMap.empty[Int, MetricsResult]
```
Comment (hvanhovell) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala]: Make this val.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala [hunk 5]
```diff
@@ -309,4 +314,94 @@ class SQLMetricsSuite extends SparkFunSuite with SharedSQLContext {
...
+      }
+      res.sumMaxOutputRows += maxOutputRows
+    }
+  }
+
+  // Run df.collect() and return aggregated metrics for each stage.
+  def run(df: DataFrame): List[(Long, Long, Long)] = {
+    val spark = df.sparkSession
+    val sparkContext = spark.sparkContext
+    val listener = new InputOutputMetricsListener()
```
Comment (hvanhovell) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala]: Use try...finally here

File: sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala [hunk 6]
```diff
@@ -309,4 +314,94 @@ class SQLMetricsSuite extends SparkFunSuite with SharedSQLContext {
...
+     *  - sum of the highest values of "number of output rows" metric for all the tasks in the stage
+     */
+    def getResults(): List[(Long, Long, Long)] = {
+      stageIdToMetricsResult.keySet.toList.sorted.map({ stageId =>
+        val res = stageIdToMetricsResult(stageId)
+        (res.recordsRead, res.shuffleRecordsRead, res.sumMaxOutputRows)})
+    }
+
+    override def onTaskEnd(taskEnd: SparkListenerTaskEnd): Unit = synchronized {
+      val res = stageIdToMetricsResult.getOrElseUpdate(taskEnd.stageId, { MetricsResult() })
```
Comment (hvanhovell) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala]: Nit remove curly braces

File: sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala [hunk 7]
```diff
@@ -309,4 +314,94 @@ class SQLMetricsSuite extends SparkFunSuite with SharedSQLContext {
...
+    assert(res2 === (150L, 0L, 150L) :: (0L, 150L, 10L) :: Nil)
+
+    withTempDir { tempDir =>
+      val dir = new File(tempDir, "pqS").getCanonicalPath
+
+      spark.range(10).write.parquet(dir)
+      spark.read.parquet(dir).createOrReplaceTempView("pqS")
+
+      val res3 = InputOutputMetricsHelper.run(
+        spark.range(0, 30).repartition(3).crossJoin(sql("select * from pqS")).repartition(2).toDF()
```
Comment (hvanhovell) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala]: This is hard to reason about. Could you add a few lines of documentation?

**Comment by hvanhovell** [comment]
LGTM - pending jenkins.

**Comment by rxin** [comment]
Merging in master.

**Comment by jaceklaskowski** [comment]
I think that the commit has left [numGeneratedRows](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala#L344) metrics off, hasn't it? (it was added in #16829)

**Comment by ala** [comment]
True. There's a couple of lines that should be removed with this change, that were left behind. numGeneratedRows should be gone.

**Comment by jaceklaskowski** [comment]
I'll have a look at this this week and send a PR unless you beat me to it :) Thanks @ala!

**Comment by ala** [comment]
Thanks @jaceklaskowski - it's already done: https://github.com/apache/spark/pull/17939

**Pull request closed without merging**